src/bandwidth.rs

```diff
@@ -0,0 +1,271 @@
+        Ok(num_bytes)
+    }
+}
+
+impl<TInner> tokio_io::AsyncRead for BandwidthConnecLogging<TInner>
+    where TInner: tokio_io::AsyncRead
+{
+}
+
+impl<TInner> Write for BandwidthConnecLogging<TInner>
+    where TInner: Write
+{
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let num_bytes = self.inner.write(buf)?;
+        self.sinks.upload.lock().inject(num_bytes);
+        Ok(num_bytes)
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl<TInner> tokio_io::AsyncWrite for BandwidthConnecLogging<TInner>
+    where TInner: tokio_io::AsyncWrite
+{
+    #[inline]
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        self.inner.shutdown()
+    }
+}
+
+/// Returns the number of seconds that have elapsed between an arbitrary EPOCH and now.
+#[inline]
+fn current_second() -> u32 {
+    lazy_static! {
+        static ref EPOCH: Instant = Instant::now();
+    }
+
+    EPOCH.elapsed().as_secs() as u32
+}
+
+/// Structure that calculates the average bandwidth over the last few seconds.
+///
+/// If you want to calculate for example both download and upload bandwidths, create two different
+/// objects.
+struct BandwidthSink {
+    /// Bytes sent over the past seconds. Contains `rolling_seconds + 1` elements. Only the first
+    /// `rolling_seconds` elements are taken into account for the average, while the last element
+    /// is the element to be inserted later.
+    bytes: SmallVec<[u64; 8]>,
+    /// Number of seconds between `EPOCH` and the moment we have last updated `bytes`.
+    latest_update: u32,
+    /// Number of seconds. Configured at initialization and never modified.
+    rolling_seconds: u32,
+}
+
+impl BandwidthSink {
+    /// Initializes a `BandwidthSink`.
+    fn new(seconds: u32) -> Self {
+        BandwidthSink {
+            bytes: smallvec![0; seconds as usize + 1],
+            latest_update: current_second(),
+            rolling_seconds: seconds,
+        }
+    }
+
+    /// Returns the number of bytes over the last few seconds. The number of seconds is the value
+    /// configured at initialization.
+    fn get(&mut self) -> u64 {
+        self.update();
+        self.bytes.iter()
+            .take(self.rolling_seconds.saturating_sub(1) as usize)
+            .fold(0u64, |a, &b| a.saturating_add(b)) / u64::from(self.rolling_seconds)
+    }
+
+    /// Notifies the `BandwidthSink` that a certain number of bytes have been transmitted at this
+    /// moment.
+    fn inject(&mut self, bytes: usize) {
+        self.update();
+        if let Some(last) = self.bytes.last_mut() {
+            *last = last.saturating_add(bytes as u64);
+        }
+    }
+
+    /// Updates the state of the `BandwidthSink` so that the last element of `bytes` contains the
+    /// current second.
+    fn update(&mut self) {
+        let current_second = current_second();
+        for _ in self.latest_update .. current_second {
```

Comment:
Maybe we should limit the range to `self.bytes.len()` to avoid unnecessary loop iterations?